### ISSUE TITLE
Wrong definition of extensions on Windows/Cygwin

### ISSUE THREAD

**Issue by braibant** [dbuenzli/topkg#11]
This commit https://github.com/dbuenzli/topkg/commit/605fc11a48da220fcef7e0293f183a4a3faf0293 introduced an issue that prevents compilation of some `topkg` packages under Cygwin on Windows. Specifically, the lines https://github.com/dbuenzli/topkg/blob/master/pkg/topkg.ml#L155-156 should test whether `Sys.cygwin` is true to select the right extention when compiling under Cygwin.

**Comment by braibant** [comment]
Edit: actually, the issue is more subtle than that. The version of OCaml we work with is the one from is compiled with the Mingw, but the compilation of packages on Windows uses Cygwin for some of the toolchain. In the latter case, the `Sys.cygwin` bool will still be false...

**Comment by dbuenzli** [comment]
Do you see any way to resolve the issue ? 

Just out of curiosity on which packages do you see the issue (I don't think I made a lot of releases that include this fix) ?

**Comment by braibant** [comment]
https://github.com/whitequark/ppx_deriving is the one. 

Not sure about how to solve this issue in a portable and clean way.  The issue is that there is no strong connection between the way the compiler was compiled (the `Sys.os_type` def.) and the tool chain that is later used in the compilation of other packages. 

One could have an extra "tool-chain" argument to the invocation of `ocaml pkg/build.ml`, but I am not sure how it would fare...

**Comment by dbuenzli** [comment]
If a new variable is needed we should coordinate this with `opam` since the next version will provide windows support. And how do the other systems `oasis`, `OCamlMakefile` handle this ?

Actually this looks like a typical cross-compilation problem, I was not paying attention when I accepted the patch. 

Only build system problems are allowed to use `Sys.*` information, build artefacts should be conditioned on an appropriate compilation *target* platform variable (I actually took great care to have that distinction in assemblage). 

Would the opam `os` variable give us the right value in that context ?

/cc @Altgr

**Comment by dbuenzli** [comment]
Btw. I guess that for `opam`, the `os` variable is also the host system. We don't have anything for the target system, now I guess that introducing variable for this goes against the idea of having compilers as packages.

**Comment by dbuenzli** [comment]
Gah also I didn't pay attention to the fact the use of `Sys.win32` implies at least `4.01`.

**Comment by braibant** [comment]
I just came back to this issue while hacking through 
https://github.com/braibant/ocaml-windows-bootstrap
https://github.com/braibant/windows-opam-repository

Since there is a fair number of packages which uses topkg on my plate, it would be nice to find a solution.

**Comment by dbuenzli** [comment]
What invocation can determine the right variable ?

**Comment by braibant** [comment]
I am not sure there is a simple invocation that can determine the right variable. It seems that mingw use `.a` as the extension for static libraries (which makes sense). So `Sys.os_type`  is not precise enough to determine the right extension here.

Would it make sense to have an extra option for `topkg` to allow the end user to override the definition of things like the extensions?

**Comment by dbuenzli** [comment]
I'd like to be sure I understand and solve this correctly. I'll soon resume my work on assemblage and hopefully `topkg` shall be deprecated in a few month, but in any case the problem will also be there (see https://github.com/samoht/assemblage/issues/130).

> So Sys.os_type is not precise enough to determine the right extension here.

Using `Sys.os_type` was anyway a very bad idea in the first place, I didn't pay attention. It is very wrong from a cross-compilation point of view, this variable gives us the host os, not the target os.

> Would it make sense to have an extra option for topkg to allow the end user to override the definition of things like the extensions?

Yes. 

Do you think that your scenario can be captured by a user settable `target-os` variable ? In your case simply setting the variable to `cygwin`, which entails unix-like suffixes would solve the problem. 

But could you forsee scenarios where e.g. you'd like still like `target-os` to be `Win32` (because it configures other things) while still having unix-like suffixes ? 

If that's the case maybe we should rather have a broad `exts` variables with `unix-like` or `win32-like` values. I think it would be cleaner and would capture your configuration scenario in a precise way.

**Comment by braibant** [comment]
I am not sure setting the variable to `cygwin` would solve the issue (but it depends what you mean with Cygwin). In my case, I am using the `mingw` toolchain which entails unix-like suffixes for some files but not other. For instance, static libraries should be '.a' while binaries are '.exe'. Using the `cygwin` port of OCaml, binaries would not use `.exe`. 

I think what make sense is to have values for the `target-os`, the toolchain used and the `host-os`, and select the extensions based on the three (with the proviso that not all combinations would make sense). The `mingw` toolchain can be used with `host-os` set to `Unix` and `target-os` set to `Win32`, for instance. 
I expect cross compilation to work the same in general. (I wonder how this is solved for android.... @whitequark might have an opinion about this).

**Comment by whitequark** [comment]
I agree wrt `target-os`. I think it's best to avoid explicit setting of extensions. With `opam-android` this was irrelevant since the host and target architectures are by and large the same--on OS X you still have `.a`, and I have never used shared libraries.

**Comment by braibant** [comment]
Well, `target-os` does not cut it by itself for Windows. We need at least the `host-os` to complement that (and adding the toolchain would be more precise). Setting explicit extensions might be a "last resort" solution...

**Comment by dbuenzli** [comment]
To be honest I'm a little bit lost now and I'm not sure I understand the problem anymore. 

`topkg` is two things:

1. A way to create the list of the products you want to build according to a build environement
2. A single invocation of your build system with this list of products.

Regarding 1. it seems that it is `topkg`'s duty to actually determine the right extensions it should build given the target platform. This should be conditioned on some kind of variable of which `target-os` seem to me to be the right one. In my poor understanding a value of `win32` for `target-os` should add `.exe` to executable, replace `.a` by `.lib` and `.so` by `.dll`; a value of `cygwin` should keep things as they are now.

Regarding 2. it should be noted that the notion of "toolchain" you refer to does not exist in the current `opam` eco-system in general and in `ocamlbuild` in particular. However we can set the individual tool to use manually using the `-ocaml*` options (and I guess only the sources know what happens whenever you use these options in conjunction with the `-use-ocamlfind` option...). It should however be decided whether it is `topkg`'s business to handle that the right tools are picked up or `ocamlbuild` or the `opam` environment via its control over the `$PATH`.

Could you maybe point to an example that does that correctly or provide some pseudo code on how this should be handled correctly. Also could you try to spell out a strategy so that we can work in hand with `opam` (even if this means adding inexisting variables in `opam`) so that in the long run we get to *platform-independent* `opam` files that do pass enough information to the build systems through `opam` variables for supporting windows/cross-compilation scenarios. 

Even if the problem is not solvable in `topkg` itself, I'm really interested to have that work in the `assemblage` context, so all this discussion should not be lost time in the long run... I'm a little bit cold on the subject at the moment but I'll soon warm up again on these problems in the upcoming weeks.

**Comment by braibant** [comment]
Maybe toolchain was not the right word to use. What I was saying is simply the following:
- using the OCaml MSVC ports, `target-os` is `win32`
- using the OCaml mingw ports, `target-os` is also `win32` 
- using the cygwin ports, `target-os` is `cygwin`

Since the extensions are not the same between the MSVC ports and the mingw ports, even if the `target-os` is the same, one need a bit more to select the extensions correctly.

**Comment by whitequark** [comment]
The notion of a toolchain actually exists, in ocamlfind, I'm using it in opam-android. You can use `ocamlfind -toolchain android` or, when this is inconvenient, e.g. with oasis or ocamlbuild, `OCAMLFIND_TOOLCHAIN=android ocamlbuild ...`.

**Comment by whitequark** [comment]
You can take a look at e.g. my Android port of tgls: https://github.com/whitequark/opam-android/blob/master/packages/tgls-android.dev/opam

**Comment by dra27** [comment]
I've just hit this compiling cmdliner. There are two ways (used in various configure-based build systems - see, for example, https://github.com/ygrek/ocurl/commit/b537a2b7259d2230b66d6aa3e104cb0e54beddc5) - you either grep the output of ocamlc/ocamlopt -config which contains the values ext_obj, ext_lib and ext_dll or, if you're happy with OCaml 4.00+ requirement, use `Config.ext_obj`, `Config.ext_lib` and `Config.ext_dll` - requires `#directory "+compiler-libs"` and `#load "ocamlcommon.cma".`

**Comment by dbuenzli** [comment]
@dra27 Using the `Config` module will not work in a cross compiler setting, besides for `cmdliner` I do need to maintain `3.12` compatibility because of the opam user. 

I'd like to solve these problems once and for all and especially in the assemblage setting I would prefer not to rely on grepping the output of `ocamlc` and hence would welcome if people could actually help me answer the questions I raised here: http://lists.ocaml.org/pipermail/wg-windows/2015-July/000032.html

**Comment by dra27** [comment]
For my own edification, why doesn't `Config` work for cross-compilation?

**Comment by dbuenzli** [comment]
Because it will give you the values for the `build-os` toolchain not for the `host-os` toolchain. See https://github.com/samoht/assemblage/issues/126#issuecomment-117189870 for more details.

**Comment by dra27** [comment]
Fair enough - although it sounds like a "bug" in cross compiling, surely? If you've got an ocamlopt cross-compiling from Linux to Windows, isn't it reasonable to expect the output of ocamlopt -config would refer to the host-os (i.e. what ocamlopt will be compiling to)? Or is that just not how cross-compilation tends to work?

**Comment by dbuenzli** [comment]
> Fair enough - although it sounds like a "bug" in cross compiling, surely?

Well no, you want the `Config` module to refer to the right toolchain, since the `topkg` script is interpreted  here on the `build-os` it should give the values right for the `build-os`.  

> If you've got an ocamlopt cross-compiling from Linux to Windows, isn't it reasonable to expect the output of ocamlopt -config would refer to the host-os (i.e. what ocamlopt will be compiling to)?

Yes, but that may not be the same compiler as the one that inteprets the `topkg` script (so it will have a different stdlib with a different `Config` module).

**Comment by dra27** [comment]
So just to see if I understand correctly - the `Config` module is behaving as I'd expect - the point is that you're running `topkg.ml` in a *different* `ocamlrun` (I'm assuming that the ocamlrun built in a cross-compiled environment only runs on host-os - i.e. a cross-compiled ocaml is designed to run on the host-os or is that not true?).
Presumably at some point `topkg.ml` is being told to invoke a different compiler from the one it's being interpreted in, so what's wrong at that point loading the correct ocamlcommon.cma for that platform and get the Config.cmo file for the host-os and the answers you need? The only requirement then is that all the installation must be the same version, I think?

**Comment by dbuenzli** [comment]
> the point is that you're running topkg.ml in a different ocamlrun

Yes.

> (I'm assuming that the ocamlrun built in a cross-compiled environment only runs on host-os - i.e. a cross-compiled ocaml is designed to run on the host-os or is that not true?).

That's my *impression* aswell. 

> so what's wrong at that point loading the correct ocamlcommon.cma for that platform and get the Config.cmo file for the host-os and the answers you need? The only requirement then is that all the installation must be the same version, I think?

That could be done, but I can see people having reasons to have different versions of the compiler in cross-compilation settings. If we don't define an `os` for `mingw` (which you hinted should not be done, still waiting for a definitive answer from you on list) I think I'm just going to parse the output of `ocamlc -config` in some way.

**Comment by dbuenzli** [comment]
Each time I try to fix that, I end up vomiting at some point.

**Comment by dbuenzli** [comment]
So the challenge is how do we invoke the right `ocamlc` so that things work in a cross compilation setting (e.g. if people [play](https://github.com/whitequark/opam-android/blob/master/packages/react-android.1.2.0/opam#L13) with the `OCAMLFIND_TOOLCHAIN` env variable). 

I want to have a solution that avoids relying solely on `ocamlfind` since some projects are using `topkg` without `ocamlfind` (e.g. `cmdliner` to be able to build `opam` without needing `ocamlfind`).

So maybe the [`builder`](https://github.com/dbuenzli/topkg/blob/a992b1ea62ba0c46765c7ba7aeb5d8aa751d4ff0/pkg/topkg.ml#L47) type should be changed from 
```ocaml
type builder = [ `OCamlbuild | `Other of string * string ]
```
to
```ocaml
type builder = 
[ `OCamlbuild of string list 
| `OCamlbuild_no_ocamlfind of string list 
| `Other of string * string ]
```

Now in each case we want to be able to do two things: 1) invoke the right `ocamlc` 2) invoke the builder. In both cases in a way that is correct w.r.t. to cross compilation. Tentative: 

1. `` `OCamlbuild args``.
    1. `ocamlc` simply invoke `ocamlfind ocamlc`
    2. `ocamlbuild` simply invoke with `ocamlbuild args -use-ocamlfind`

    This should play nice with `OCAMLFIND_TOOLCHAIN` hack.
2. `` `OCamlbuild_no_ocamlfind args``. We assume we have an environment variable `HOST_OS_XBIN` or a command line argument to `topkg` that holds a path to the directory that has the cross compilation toolchain binaries.
    1. `ocamlc` is invoked as `$HOST_OS_XBIN/ocamlc` or `ocamlc` if the variable is empty.
    2. `ocamlbuild` is invoked as `ocamlbuild` if `$HOST_OS_XBIN` is empty and otherwise painfully as:
    ``` 
ocamlbuild args -ocamlc $HOST_OS_XBIN/ocamlc$(opt) 
                    -ocamlopt $HOST_OS_XBIN/ocamlopt$(opt) 
                    -ocamldep $HOST_OS_XBIN/ocamldep$(opt) 
                    -ocamldoc $HOST_OS_XBIN/ocamldoc$(opt) 
                    -ocamlyacc $HOST_OS_XBIN/ocamlyacc$(opt) 
                    -menhir $HOST_OS_XBIN/menhir$(opt)
                    -ocamllex $HOST_OS_XBIN/ocamllex$(opt)
                    -ocamlmklib $HOST_OS_XBIN/ocamlmklib$(opt)
                    -ocamlmktop $HOST_OS_XBIN/ocamlmktop$(opt)
                    -ocamlrun $HOST_OS_XBIN/ocamlrun$(opt) 
    ```
with `$(opt)` equal to `".opt"` if `native-tools` is `true` or `""` otherwise. With `native-tools` a new argument given to `topkg` which should be given by `opam`'s `ocaml-native-tools` variable (assuming in the future it will be able to give us the right information for the toolchain being used). Well maybe it's better not to try to handle the `.opt` thing it only introduces more potential for failure and packages using this case should be rare so we can pay the price of a build with byte code tools (I also filed an [issue](http://caml.inria.fr/mantis/view.php?id=7014) upstream, it seems to me that these `.opt` binaries are only there to make our build life more miserable)

2. `` `Other (build, _)``, we assume we have the same `$HOST_OS_XBIN` variable or argument.
   1. `ocamlc` is invoked as `$HOST_OS_XBIN/ocamlc`
   2. The builder is invoked as `build`.

@whitequark If you have time to read this I welcome your comments.

**Comment by whitequark** [comment]
This looks reasonable to me.

**Comment by dbuenzli** [comment]
Thanks @whitequark.

I'd like to definitively fix the windows case once for all (hopefully). It seems that `ocamlc -config` will not give me the extension for executables. So I'm simply going to extract the `os_type:` and `ccomp_type:` fields of that output and apply the following reasoning (derived from this [message](http://lists.ocaml.org/pipermail/wg-windows/2015-July/000037.html)): 

```ocaml
let ext_obj, ext_lib, ext_dll, ext_exe = match ccomp_type, os_type with
| "msvc", _     -> ".obj", ".lib", ".dll", ".exe"
| "cc", "Win32" -> ".o",   ".a",   ".dll", ".exe"
| _, _          -> ".o",   ".a",   ".so",  ""
```

Now the only remaining problem is how to apply the `ext_exe` to binaries on windows in the various cases. What does `ocamlbuild` generate when you do `ocamlbuild exec.{byte,native}` ? does it generate `_build/exec.{byte,native}` or `_build/exec.{byte,native}.exe` and this in every Windows case ?

**Comment by whitequark** [comment]
Just `.native`, and looking at ocamlbuild's Makefile, which contains things like `ocamlbuild.native: ocamlbuild_pack.cmx $(EXTRA_CMX) ocamlbuild.cmx`, I question whether anyone has ever built ocamlbuild on Windows.

**Comment by whitequark** [comment]
Reading further, it does `$(CP) ocamlbuild.native $(INSTALL_BINDIR)/ocamlbuild$(EXE)` during installation. It also has `Ocamlbuild_config.exe` with the corresponding extension, but it is only used while building a plugin. I feel that ocamlbuild might have to be fixed first...

**Comment by dbuenzli** [comment]
Let's wait for @dra27 to chime in. I vaguely remember he might have fixed ocamlbuild issues on its ongoing work to port opam to windows.

**Comment by dra27** [comment]
ocamlbuild indeed *doesn't* add `.exe` to executable outputs on Windows - it's left to the user to rename the output. That's never felt particularly weird to me, since one would usually want to rename at install-time to get rid of .native or .byte anyway, although it is arguably inconsistent that you can't run foo.native or foo.byte directly.

ocamlbuild does work in a suitable environment (it needs access to sh), though personally I've never used it directly - only as part of other libraries, and typically wrapped by OASIS (which I'm fairly sure handles the renaming for .exe itself).

**Comment by dbuenzli** [comment]
@dra27 thanks for the input. I now have something in, but it will need testing before I start propagating this; I'll wait on next week's meeting to see if you maybe got to a state so that I can directly test this myself.

**Comment by whitequark** [comment]
@dbuenzli Any news? I'm getting reports like https://github.com/whitequark/ppx_import/issues/6...

**Comment by dbuenzli** [comment]
I merged the support in `master` and started to distribute packages that use this but I never managed to properly test it.

**Comment by whitequark** [comment]
@dbuenzli It's broken unfortunately: https://github.com/whitequark/ppx_import/issues/6#issuecomment-165692021

**Comment by dbuenzli** [comment]
So on Windows with ocamlbuild you still do `ocamlbuild exec.native` to build an executable ?

**Comment by whitequark** [comment]
Yep.

**Comment by whitequark** [comment]
And yes, you cannot immediately run it, which is dumb, but that's what we have to work with, I guess...

**Comment by whitequark** [comment]
I filed this as https://github.com/ocaml/ocamlbuild/issues/6.

**Comment by dra27** [comment]
And yes, you cannot immediately run it, which is dumb, but that's what we have to work with, I guess...

It's not ideal, but it's also not necessarily worth the overhead fixing at the ocamlbuild end - beyond debugging (or for a tool used in a build process), how often do you actually install a .byte or .native file without renaming it?

**Comment by whitequark** [comment]
I run .byte or .native files all the time during development. Moreover, you are shipping a tool which is broken by default and hostile to any novice developer.

**Comment by dra27** [comment]
I think there are waay many other things in the way of novice users of ocamlbuild :wink: However, for a developer, I still think the fairly trivial reconfiguration of one's computer to execute .byte and .native files is preferable to adding .exe - given the already weird inclusion of a build instruction in the target name already and the cross-platform confusion it generates here for less gain.

**Comment by dra27** [comment]
Hang on a minute, I was cheating answering from my iPhone. No reconfiguration should be necessary - what do you mean when you say you can't run the .native and .byte files?

**Comment by whitequark** [comment]
> However, for a developer, I still think the fairly trivial reconfiguration of one's computer to execute .byte and .native files

Are you serious? This is a ridiculous suggestion. No other build tool does this and there is no reason for ocamlbuild to not be fixed except for some irrational resistance to change.

As for extensions, this already happened with .a/.lib.

**Comment by whitequark** [comment]
Windows does not consider .byte/.native files executable.

**Comment by dra27** [comment]
The shell most certainly does - you mean you run .byte and .native files from Windows Explorer with a mouse as part of your development process?

**Comment by whitequark** [comment]
The *Cygwin* shell. cmd does not. Anything that does not use cygwin, e.g. other OCaml tools that use MinGW, do not. Explorer does not, and yes, this is how I would run a GUI app if I ever had to develop one on Windows (which has not happened so far).

That's all beside the point though, because there is no valid reason for this egregious violation of the target platform's conventions.

**Comment by dra27** [comment]
What version of Windows are you on? It certainly does work here! If I add .native, I don't even have to type the extension (if I so desired).
It's also not necessarily a violation, which you seem to be viewing it as. The ocamlbuild specification is just a build target instruction - it's just a side-effect that its output is also executable.

**Comment by whitequark** [comment]
> It certainly does work here!

What OCaml distribution and Windows version are you using?

> It's also not necessarily a violation, which you seem to be viewing it as. The ocamlbuild specification is just a build target instruction - it's just a side-effect that its output is also executable.

The build target does conform to platform conventions for basically everything else--object files, static or shared libraries, etc.

**Comment by dra27** [comment]
Yes, but in those cases it's building actual files in their final names. I'm really not convinced that the extension to executables (which can and in virtually all cases will be renamed) is a given and, as I say, even desirable - given that suddenly the way in which you tell ocamlbuild to compile programs becomes platform-specific.
The machine I just tried it on is Windows 7 with OCaml 4.01.0 MinGW - I'm wondering if you're confusing the potentially broken symlinks which ocamlbuild places in the output directory with the lack of a .exe extension? The .native and .byte files in the _build directory should most certainly be directly runnable in cmd.
It's been possible since the days of MS-DOS to name an executable file with an extension other than .exe - you just have to specify the extension if it's not .exe (or, these days, included in `PATHEXT`)

**Comment by whitequark** [comment]
> given that suddenly the way in which you tell ocamlbuild to compile programs becomes platform-specific.

It is *already* platform-specific. topkg uses the platform to figure out how to build a library, for example.

> The .native and .byte files in the _build directory should most certainly be directly runnable in cmd.

Let me verify. Regardless, it is still a problem if they cannot be run with Explorer.

**Comment by dra27** [comment]
I am aware of libraries - my attempt at a point is that you have no choice but name those files correctly - you do have a choice with the executable, and it's really not clear that it's necessarily a good one to go for a platform-specific extension. Library extensions (.lib, .a, .obj, .o, etc.) are imposed by things outside of ocamlbuild's control - .byte and .native are ocamlbuild's **own** extensions, so it can do what it likes, if it works (which it does). Conversely, it's a problem the compiler *has* to work around because the one exception to things you cannot run directly is files with *no* extension - so `ocamlopt -o foo foo.ml` would result in an actually unexecutable file `foo` which is why it must append `.exe`. But you can't invoke ocamlbuild without appending a target extension.
Why is it a problem if they cannot be run in Explorer (and they can - having to configure a file type is really not a major thing - indeed, it's something which ocaml in a cuddly setup *should* be doing to allow .ml files to be executed from Explorer, just as ActivePerl, ActiveTcl and every other Windows distribution of an interpreted language does)? But it's hardly critical: they are console applications, unless you're considerably beyond an OCaml beginner, and they have a .byte and .native in the filename which should never see the light of release anyway!

**Comment by dbuenzli** [comment]
Couldn't get a confirmation that it works but I'll assume it does. Windows user please open a new issue if it does not.
